core/src/main/java/io/grpc/internal/ManagedChannelOrphanWrapper.java

```diff
@@ -0,0 +1,163 @@
+    private static final RuntimeException missingCallSite = missingCallSite();
+
+    private final ReferenceQueue<ManagedChannelOrphanWrapper> refqueue;
+    private final ConcurrentMap<ManagedChannelReference, ManagedChannelReference> refs;
+
+    private final ManagedChannel channel;
+    private final Reference<RuntimeException> allocationSite;
+    private volatile boolean shutdown;
+    private volatile boolean shutdownNow;
+
+    ManagedChannelReference(
+        ManagedChannelOrphanWrapper orphanable,
+        ManagedChannel channel,
+        ReferenceQueue<ManagedChannelOrphanWrapper> refqueue,
+        ConcurrentMap<ManagedChannelReference, ManagedChannelReference> refs) {
+      super(orphanable, refqueue);
+      allocationSite = new SoftReference<RuntimeException>(
+          ENABLE_ALLOCATION_TRACKING
+              ? new RuntimeException("ManagedChannel allocation site")
+              : missingCallSite);
+      this.channel = channel;
+      this.refqueue = refqueue;
+      this.refs = refs;
+      this.refs.put(this, this);
+      cleanQueue(refqueue);
+    }
+
+    /**
+     * This clear() is *not* called automatically by the JVM.  As this is a weak ref, the reference
+     * will be cleared automatically by the JVM, but will not be removed from {@link #refs}.
+     * We do it here to avoid this ending up on the reference queue.
+     */
+    @Override
+    public void clear() {
+      clearInternal();
+      // We run this here to periodically clean up the queue if at least some of the channels are
+      // being shutdown properly.
+      cleanQueue(refqueue);
+    }
+
+    // avoid reentrancy
+    private void clearInternal() {
+      super.clear();
+      refs.remove(this);
+      allocationSite.clear();
+    }
+
+    private static RuntimeException missingCallSite() {
+      RuntimeException e = new RuntimeException(
+          "ManagedChannel allocation site not recorded.  Set -D"
+              + ALLOCATION_SITE_PROPERTY_NAME + "=true to enable it");
+      e.setStackTrace(new StackTraceElement[0]);
+      return e;
+    }
+
+    @VisibleForTesting
+    static int cleanQueue(ReferenceQueue<ManagedChannelOrphanWrapper> refqueue) {
+      ManagedChannelReference ref;
+      int orphanedChannels = 0;
+      while ((ref = (ManagedChannelReference) refqueue.poll()) != null) {
+        RuntimeException maybeAllocationSite = ref.allocationSite.get();
+        ref.clearInternal(); // technically the reference is gone already.
+        if (!(ref.shutdown && ref.channel.isTerminated())) {
+          orphanedChannels++;
+          Level level = ref.shutdownNow ? Level.FINE : Level.SEVERE;
+          if (logger.isLoggable(level)) {
+            String fmt = new StringBuilder()
+                .append("*~*~*~ Channel {0} was not ")
+                // Prefer to complain about shutdown if neither has been called.
+                .append(!ref.shutdownNow ? "shutdown" : "terminated")
```

Comment:
Why was `ref.shutdown` changed to `ref.shutdownNow`?

Reply:
Fixed, this was my mistake when fixing variables with the IDE